examples/scan_iterator.py

```diff
@@ -0,0 +1,38 @@
+"""
+Scan/Discovery Async Iterator
+--------------
+
+Example showing how to scan for BLE devices using async iterator instead of callback function
+
+Created on 2023-07-07 by bojanpotocnik <info@bojanpotocnik.com>
+
+"""
+import asyncio
+
+from bleak import BleakScanner
+
+
+async def main():
+    x = 5
+    print(f"Scanning for {x} devices...")
+    async for bd, ad in BleakScanner.scan():
+        print(f"{x}. {bd!r} with {ad!r}")
+        x -= 1
+        if x == 0:
+            break
+
+    # Ensure that asyncio loop has time to stop the scanner via the context manager exit
+    await asyncio.sleep(0.1)
```

Comment:
Ugh. I forgot about the issue with unfinished generators hanging around without cleaning up. So it is probably best to just avoid the issue and leave out the `scan()` method.

Reply:
The alternative is to always write like this instead:

```python
    print(f"Scanning for {x} devices...")
    dev_gen = BleakScanner.scan()

    try:
        async for bd, ad in dev_gen:
            print(f"{x}. {bd!r} with {ad!r}")
            x -= 1
            if x == 0:
                break
    finally:
        # Ensure that the scanner is stopped
        await dev_gen.aclose()
```

Reply:
I understand – thank you. This then diminishes the whole point of having a class method, as it is even more work than using instance method :smile:

I now solved it by changing the order of examples – this solves the issue. I can still remove the `scan()` method if desired (would also fit my long comment above https://github.com/hbldh/bleak/pull/1361#discussion_r1258890610 )

Reply:
> This then diminishes the whole point of having a class method

I very much agree and think we should leave it out.

> I now solved it by changing the order of examples

I think this just hides the problem rather than solving it.

Reply:
For the future archeologists, here is perhaps a better way to do a classmethod async iterator with deterministic cleanup (but still not very nice).

```python
    print(f"Scanning for {x} devices...")
    
    async with contextlib.aclosing(BleakScanner.scan()) as dev_gen:
        async for bd, ad in dev_gen:
            print(f"{x}. {bd!r} with {ad!r}")
            x -= 1
            if x == 0:
                break
```

Reply:
This is added in the example https://github.com/hbldh/bleak/pull/1361/files#diff-547d9f54fbc8eafa2aa817c1d50d8e7cfd3958c4f5983d0923bf79a99b255222R37
I hope it is well enough separated to not confuse regular users while providing useful insight in async resource cleanup.

I am afraid nobody will find this comments here, or at least will took them too much time.